(Feature) Issue #47 add logic that uses the watch-depth HTML attribute t...

djett41 opened the pull request:
(Feature) Issue #47 add logic that uses the watch-depth HTML attribute to determine how to watch props on scope with the following work flow.  

watchDepth equal to 'collection' ? $watchCollection is invoked
watchDepth equal to 'group' ? $watchGroup is invoked
watchDepth equal to 'reference' ? $watch is invoked (no deep watch)
otherwise $watch is invoked with deep watching by value (true sent in as 3rd argument)

kasperp commented:
@djett41 :+1: Thanks it looks good. 

About your questions:

> .. wasnt sure if you wanted to minify yourself in case you had other pull requests coming,

You can commit the minified file as well.

> Also I'm not too familiar with coffee script (assuming thats what your tests are in?) so I didnt add to your test file :( but let me know if you would like me to create a separate test file just for this feature.

The tests are written in javascript, we just run the reactify with the es6 option. Which gives us some es6 features like arrow functions, and yes it would be nice if you wrote some tests :smile: 

One other thing that sprung to mind:  which versions of angular this fix requires? so far we had no restrictions, but `watchGroup` is a 1.3? feature and `watchCollection` is a 1.1.4? feature. Would it be possible to "feature detect" instead of breaking if these functions are not available ?

If you'd like to add this description of this feature to the README.md that would also be nice, but I can also do that :smile:

djett41 commented:
haha please ignore my ignorance ;-) when I saw the arrows my brain thought coffeescript. 

Sounds good, and good catch, as I didnt take angular versions into consideration.  I'll take care of these items and send you another pull request.  I can add to the README as well.

kasperp commented:
Cool thanks again

djett41 commented:
@kasperp I added the feature detection logic, added tests, updated README and built the minified file.

There are however 2 changes.  
1.  watchGroup doesn't really make sense in ngReact since we are watching either the props object (reactComponent) or the individual properties of props (reactDirective).  There will never be a prop that is an expression of other props, so it doesn't really make sense.

2.  Since reactComponent does a watch on the whole evaluated props object, rather than watching each property individually, watchCollection doesn't make sense for reactComponent as the value of props can't be an array.  for 'collection' to be useful in reactComponent, we would have to loop through keys in props and watch each property individually as reactDirective does. 

The readme specifies which watch-depth values you should use for reactComponent and reactDirective

kasperp commented on README.md:
````diff
@@ -124,12 +125,12 @@ This creates a directive that can be used like this:
 ```
 
-The `reactDirective` service will read the React component `propTypes` and watch attributes with these names. If your react component doesn't have `propTypes` defined you can pass in an array of attribute names to watch.
+The `reactDirective` service will read the React component `propTypes` and watch attributes with these names. If your react component doesn't have `propTypes` defined you can pass in an array of attribute names to watch.  By default, attributes will be watched by value however you can also choose to watch by reference or collection by supplying the watch-depth attribute.  Possible values are `ref ...
````
no longer collection - right ?

kasperp commented on ngReact.js:
```diff
@@ -82,6 +113,31 @@
+    var watchFn;
+
+    //default watchDepth to value if not reference or collection
+    if (isWatchCollectionAvailable && watchDepth === 'collection') {
```
Would a simpler `if ( angular.isFunction( '$watchCollection' && .. ` not be sufficient. The nice thing about this approach is that the detection doesn't happen based on an exact version.

djett41 replied:
The isFeatureAvailable function would compare min for that feature against current angular version so min of 1.3.12 for a feature against 1.4 would return true.

However you're right, checking angular.isFunction(scope.$watchCollections) is alot simpler!  Thanks for pointing that out, I made the change in ngReact,js and minified the file

kasperp commented on tests/react-directive-spec.js:
```diff
@@ -25,6 +25,22 @@ var Hello = React.createClass({
+    var names = people.map(function(person, index) {
+      var name = person.fname + ' ' + person.lname;
+      return (++index === people.length) ? name : name + ', ';
+    });
```
Could you use `people.join( ', ' )` instead of the map ?

djett41 replied:
Since items in the array are objects and i'm concatenating person.fname and person.lname I can't just use join, however I did update the code to use join for the name printing rather than using index.  Thanks!

kasperp commented:
@djett41 Looks great. I added a couple of minor things. Regarding the feature detection is there any reason, I'm missing, why we would do it based on version ?

With those taken care of I think we can merge it, @davidchang any comments before I do so ?

kasperp commented on ngReact.js:
```diff
@@ -82,6 +82,31 @@
+    var watchFn;
+
+    //default watchDepth to value if not reference or collection
+    if (watchDepth === 'collection' && angular.isFunction(scope.$watchCollections)) {
```
shouldn't it be `angular.isFunction(scope.$watchCollection)` (no 's')

kasperp commented on tests/react-directive-spec.js:
```diff
@@ -25,6 +25,20 @@ var Hello = React.createClass({
+  render(){
+    var names = this.props.items.map(function(person){
+      return person.fname + ' ' + person.lname;
+    }).join(', ');
```
:+1: you could use a arrow function `.map(p => p.fname  + ' ' p.lname).join(',')`

djett41 commented:
thanks for the catch @kasperp updated the files once again.  since the 'else' defaults to watching by value the tests still passed :-/

kasperp commented:
@davidchang what you think - new release ?
